toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -110,13 +119,85 @@ func fetchPackages() (err error) {
 		return
 	}
 
+	// Optional delta build cache hydration
+	if *tryDownloadDeltaRPMs {
+		logger.Log.Info("Attempting to download delta RPMs for build nodes")
+		err = downloadDeltaNodes(dependencyGraph, cloner)
+		if err != nil {
+			err = fmt.Errorf("failed to download delta RPMs: %w", err)
+			return
+		}
+		// Update the package graph with the paths to the delta RPMs we downloaded
+		err = pkggraph.WriteDOTGraphFile(dependencyGraph, *outputGraph)
+		if err != nil {
+			err = fmt.Errorf("failed to write cache graph to file: %w", err)
+			return
+		}
+	}
+
 	// If we grabbed any RPMs, we need to convert them into a local repo
-	if hasUnresolvedNodes {
+	if *tryDownloadDeltaRPMs || hasUnresolvedNodes {
+		logger.Log.Info("Configuring downloaded RPMs as a local repository")
 		err = cloner.ConvertDownloadedPackagesIntoRepo()
 		if err != nil {
 			err = fmt.Errorf("failed to convert downloaded RPMs into a repo:\n%w", err)
 			return err
 		}
+
+		if strings.TrimSpace(*outputSummaryFile) != "" {
+			err = repoutils.SaveClonedRepoContents(cloner, *outputSummaryFile)
+			if err != nil {
+				logger.Log.Errorf("Failed to save cloned repo contents.")
+				return
+			}
+		}
+	}
+	return
+}
+
+// downloadDeltaNodes will look at the final cached graph we saved and see if any RPMS can be download instead of built.
+// If the previous part of the fetcher worked well we should be able to download only the delta RPMs we need
+// to build our packages or image (i.e. we should be able to create a subgraph just like we would for the build step)
+//   - dependencyGraph: The graph to use to find the packages we need to build. Should have any caching operations already
+//     performed on it. Will be updated with the paths to the delta RPMs we download.
+//   - cloner: The cloner to use to download the RPMs
+func downloadDeltaNodes(dependencyGraph *pkggraph.PkgGraph, cloner *rpmrepocloner.RpmRepoCloner) (err error) {
+	const (
+		deltaBuildGraph = true
+	)
+
+	timestamp.StartEvent("delta package download", nil)
+	defer timestamp.StopEvent(nil)
+
+	// Generate the list of packages that need to be built. If none are requested then all packages will be built. We
+	// don't care about explicit rebuilds here since we are going to rebuild them anyway.
+	packageVersToBuild, err := parseAndGeneratePackageList(dependencyGraph, *pkgsToBuild, *pkgsToRebuild, *pkgsToIgnore, *imageConfig, *baseDirPath)
+	if err != nil {
+		err = fmt.Errorf("unable to generate package build list to calculate delta downloads: %w", err)
+		return
+	}
+
+	timestamp.StartEvent("graph initialization", nil)
+	// The scheduler utils expect to pick a graph up from a file, so we will write the graph we wrote it to a file and
+	// now we read it back in and optimize it. We will heavily modify this graph so it should not be used for anything
+	// else.
+	isGraphOptimized, deltaPkgGraphCopy, _, err := schedulerutils.InitializeGraph(*outputGraph, packageVersToBuild, deltaBuildGraph)
```

Comment:
>schedulerutils.InitializeGraph(*outputGraph, packageVersToBuild, deltaBuildGraph) [](http://example.com/codeflow?start=48&length=81)

I think you mentioned it during one of our discussions and alredy said will take another look at this argument, but just in case I'm thinking about something else: does the `deltaBuild` argument of `InitializeGraph` should be set to `true` with the new understanding of the delta builds we're introducing here? I think you mentioned that you're planning to re-name this argument to avoid confusion between the old and new understanding of a delta build while we're in the transition phase.

Reply:
Updated the meaning of these flags. It now is clear that we only use this to optimize the graph during the first pass in the scheduler (which is what we want here).